File: smoke_tests/php/xxe.php
<?php
// XXE (XML External Entity) vulnerabilities in PHP

// Test 1: SimpleXML with external entities
function parse_simplexml() {
    $xml = $_POST['xml'];
    // VULNERABLE: External entities enabled by default in older PHP
    $doc = simplexml_load_string($xml);
    echo $doc->asXML();
}

// Test 2: DOMDocument with DTD loading
function parse_domdocument() {
    $xml = $_POST['xml'];
    $doc = new DOMDocument();
    // VULNERABLE: Enabling external entity loading
    $doc->loadXML($xml, LIBXML_DTDLOAD | LIBXML_NOENT);
    echo $doc->saveXML();
}

// Test 3: XMLReader with external entities
function parse_xmlreader() {
    $xml = $_POST['xml'];
    $reader = new XMLReader();
    // VULNERABLE: Setting entity substitution
    $reader->setParserProperty(XMLReader::SUBST_ENTITIES, true);
    $reader->XML($xml);
    while ($reader->read()) {
        echo $reader->readString();
    }
}

// Test 4: LIBXML options vulnerable
function parse_libxml_options() {
    $xml = $_POST['xml'];
    $options = LIBXML_NOENT | LIBXML_DTDLOAD | LIBXML_DTDATTR;
    // VULNERABLE: Multiple dangerous options
    $doc = new DOMDocument();
    $doc->loadXML($xml, $options);
    echo $doc->documentElement->textContent;
}

// Test 5: Loading XML from file
function parse_from_file() {
    $filename = $_GET['file'];
    $doc = new DOMDocument();
    // VULNERABLE: Loading potentially malicious XML
    $doc->load($filename, LIBXML_NOENT);
    echo $doc->saveXML();
}

// Test 6: SOAP with XML payload
function soap_xxe() {
    $xml = file_get_contents('php://input');
    // VULNERABLE: SOAP XML parsing
    $doc = new DOMDocument();
    $doc->loadXML($xml, LIBXML_NOENT);
    // Process SOAP...
}

// Test 7: XPath with XXE document
function xpath_xxe() {
    $xml = $_POST['xml'];
    $doc = new DOMDocument();
    // VULNERABLE: XXE before XPath
    $doc->loadXML($xml, LIBXML_NOENT);
    $xpath = new DOMXPath($doc);
    $results = $xpath->query('//data');
    foreach ($results as $node) {
        echo $node->textContent;
    }
}

// Test 8: simplexml_load_file
function load_xml_file() {
    $url = $_GET['url'];
    // VULNERABLE: Loading XML from URL
    $xml = simplexml_load_file($url, null, LIBXML_NOENT);
    echo $xml->asXML();
}

// Test 9: XML stream wrapper
function xml_stream() {
    $uri = $_GET['uri'];
    // VULNERABLE: php://filter or other wrappers
    $xml = file_get_contents($uri);
    $doc = simplexml_load_string($xml);
    echo $doc->asXML();
}

// Test 10: XInclude processing
function xinclude_process() {
    $xml = $_POST['xml'];
    $doc = new DOMDocument();
    $doc->loadXML($xml);
    // VULNERABLE: XInclude can include external files
    $doc->xinclude();
    echo $doc->saveXML();
}

// Test 11: Entity expansion attack (billion laughs)
function parse_no_limits() {
    $xml = $_POST['xml'];
    $doc = new DOMDocument();
    // VULNERABLE: No entity expansion limits
    libxml_disable_entity_loader(false);
    $doc->loadXML($xml, LIBXML_NOENT | LIBXML_PARSEHUGE);
    echo $doc->saveXML();
}
?>

File: smoke_tests/cpp/xxe.cpp
// XXE (XML External Entity) vulnerabilities in C++
#include <iostream>
#include <string>
#include <libxml/parser.h>
#include <libxml/tree.h>
#include <xercesc/parsers/XercesDOMParser.hpp>

using namespace xercesc;

// Test 1: libxml2 with default settings (external entities enabled)
void parse_xml_unsafe(const std::string& xml_content) {
    // VULNERABLE: External entities enabled by default
    xmlDocPtr doc = xmlParseMemory(xml_content.c_str(), xml_content.size());
    if (doc) {
        xmlFreeDoc(doc);
    }
}

// Test 2: Loading external DTD
void parse_with_dtd(const std::string& filename) {
    // VULNERABLE: Loading external DTD
    xmlDocPtr doc = xmlReadFile(filename.c_str(), nullptr, XML_PARSE_DTDLOAD);
    if (doc) {
        xmlFreeDoc(doc);
    }
}

// Test 3: Entity expansion enabled
void parse_with_entities(const std::string& xml) {
    // VULNERABLE: NOENT causes entity expansion
    xmlDocPtr doc = xmlReadMemory(xml.c_str(), xml.size(),
                                   nullptr, nullptr, XML_PARSE_NOENT);
    if (doc) {
        xmlFreeDoc(doc);
    }
}

// Test 4: Xerces-C with external entities
void parse_xerces_unsafe(const std::string& filename) {
    XMLPlatformUtils::Initialize();

    XercesDOMParser* parser = new XercesDOMParser();
    // VULNERABLE: Not disabling external entities
    // parser->setDoNamespaces(true);
    // parser->setDisableDefaultEntityResolution(true);  // NOT SET

    parser->parse(filename.c_str());

    delete parser;
    XMLPlatformUtils::Terminate();
}

// Test 5: XInclude processing
void parse_with_xinclude(const std::string& filename) {
    xmlDocPtr doc = xmlReadFile(filename.c_str(), nullptr, 0);
    if (doc) {
        // VULNERABLE: XInclude can include external files
        xmlXIncludeProcess(doc);
        xmlFreeDoc(doc);
    }
}

// Test 6: Parameter entity processing
void parse_parameter_entities(const std::string& xml) {
    // VULNERABLE: Parameter entities can exfiltrate data
    xmlDocPtr doc = xmlReadMemory(xml.c_str(), xml.size(),
                                   nullptr, nullptr,
                                   XML_PARSE_DTDATTR | XML_PARSE_NOENT);
    if (doc) {
        xmlFreeDoc(doc);
    }
}

// Test 7: No entity expansion limits
void parse_no_limits(const std::string& xml) {
    // VULNERABLE: Billion laughs attack possible
    xmlDocPtr doc = xmlReadMemory(xml.c_str(), xml.size(),
                                   nullptr, nullptr,
                                   XML_PARSE_NOENT | XML_PARSE_DTDLOAD | XML_PARSE_HUGE);
    if (doc) {
        xmlFreeDoc(doc);
    }
}

// Test 8: XSLT with document() function
void apply_xslt_unsafe(const std::string& xml_file, const std::string& xslt_file) {
    // VULNERABLE: XSLT document() function can read local files
    xmlDocPtr doc = xmlParseFile(xml_file.c_str());
    xmlDocPtr style = xmlParseFile(xslt_file.c_str());
    // xsltStylesheetPtr xslt = xsltParseStylesheetDoc(style);
    // Apply transform...
    if (doc) xmlFreeDoc(doc);
    if (style) xmlFreeDoc(style);
}
